Musca board/soc port

galak opened the pull request:
Add support for the ARM Musca SoC/Board.  The Musca implements a dual core SoC w/ Cortex-M33 cores.

Commit:
arm: v2m musca: Add support for v2m musca SoC and board

Support musca board as a variant, it is a Cortex-M33 based SoC.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>
Signed-off-by: Minos Galanakis <minos.galanakis@arm.com>
Signed-off-by: Andy Gross <andy.gross@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

Commit:
serial: Add Support for ARM PL011 UART

This patch adds a serial driver for the ARM PL011 IP block.

Signed-off-by: Andy Gross <andy.gross@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
Need to add board docs.

agross-linaro commented on arch/arm/soc/arm/musca_a/Kconfig.series:
```diff
@@ -0,0 +1,14 @@
+	depends on ARM
+	select SOC_FAMILY_ARM
+	select CPU_HAS_SYSTICK
+	select BUILD_OUTPUT_HEX
```
this doesn't appear to be required for musca.  I can load up .bin

agross-linaro commented on boards/arm/v2m_musca/v2m_musca-nonsecure.dtsi:
```diff
@@ -0,0 +1,59 @@
+/ {
+	flash0: flash@230400 {
+		/* QSPI flash */
+		reg = <0x00230400 0x1ffc00>;
```
This needs to describe the whole part from the non-secure view.  So it'd be <0x00200000 0x800000>.  We need to use partitions to define the application partition.  for non-secure this would be the correct offset to the address which would be 0x000230400.

ioannisg replied:
So, @galak, @agross-linaro, should we agree to do this in the Board DTS?

agross-linaro commented on boards/arm/v2m_musca/v2m_musca-nonsecure.dtsi:
```diff
@@ -0,0 +1,59 @@
+
+	sram0: memory@20010000 {
+		compatible = "mmio-sram";
+		reg = <0x20010000 0x10000>;
```
Same comment here, really.  The whole sram needs to be described.  We need an application offset for this to point to the correct memory partition.  So this reg should be <0x20000000 0x20000>.  And we need something to express the application offset.  Maybe reserved-memory?

ioannisg replied:
I agree with @agross-linaro: .dtsi should describe the whole flash and sram of the SOC. However, **somewhere** (in DTS, ideally)  we need to define which will be the SRAM to be used by the NONSECURE firmware image. 

So we need something similar to the flash partitioning, where we "choose" slot-1 instead of all-flash, when we have MCUboot present.

agross-linaro commented on boards/arm/v2m_musca/v2m_musca.dts:
```diff
@@ -0,0 +1,32 @@
+
+#include <arm/armv8-m.dtsi>
+
+#include "v2m_musca-secure.dtsi"
```
we can steer the secure/non-secure here by ifdef'ing this with CONFIG_ARM_NONSECURE_FIRMWARE to select the non-secure one.

ioannisg replied:
Agree.

agross-oss reviewed (commented):
minor stuff..... but would like some thoughts on the addressing of the memory and flash for non-secure

agross-linaro commented:
Minos is to provide board documents.  We can fold those in when he gets them in to us.

ioannisg commented on arch/arm/soc/arm/musca_a/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,18 @@
+	int
+	default 50000000
+
+source "arch/arm/soc/arm/musca_a/Kconfig.defconfig.musca_a"
```
gsource, perhaps?

galak replied:
why gsource?

ioannisg replied:
Right, my mistake.

ioannisg commented on arch/arm/soc/arm/musca_a/Kconfig.series:
```diff
@@ -0,0 +1,14 @@
+
+config SOC_SERIES_MUSCA
+	bool "ARM v2m MUSCA MCU Series"
+	depends on ARM
```
`depends on ARM `is not, really, used in other SOC series definitions (other than ARM SOCs)

galak replied:
fixed

ioannisg commented on arch/arm/soc/arm/musca_a/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+
+config SOC_V2M_MUSCA_A
+	bool "ARM Cortex-M33 SMM-SSE-200 on V2M-MUSCA"
+	select CPU_CORTEX_M33
```
Just wondering; why this could not go to the SERIES definition?

ioannisg replied:
Still wondering about this, @galak

ioannisg reviewed (commented):
Some comments from my side

galak commented:
@ioannisg can you rereview.  @dbkinder has docs now, can you take a look.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+=================
+
+The v2m_musca board provides the following user push buttons:
+• PBON power on/off.
```
Use either a ``-`` or ``*`` for a bullet list (not this UTF-8 character).  Also, add a blank line before the first item in the list.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+See the `V2M Musca Website`_ for a complete list of V2M Musca board hardware
+features.
+
+The default configuration can be found in the defconfig file:
```
Delete the code-block and simply write this as:

    The default configuration can be found in the defconfig file:
    ``boards/arm/v2m_musca/v2m_musca_defconfig``.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+A Cortex-M33-based board uses vectored exceptions. This means each exception
+calls a handler directly from the vector table.
+
+Handlers are provided for exceptions 1-7, 11-12, and 14-15. The table here
```
Who provides these handlers?  Should this read:

    Zephyr provides handlers for exceptions 1-7, 11, 12, 14, and 15, as 
    listed in this table:

MaureenHelm replied:
Zephyr handles these exceptions for all Cortex M boards, why are we calling this out specifically on this board?

minosgalanakis replied:
@MaureenHelm This was added for consistency purposes, since it was included in the Beetle board documentation.It is also good to list the Secure-fault exception which is new to the v8m processors.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+|      |            | attempt to ARM |                          |
+|      |            | mode           |                          |
++------+------------+----------------+--------------------------+
+|  7   |SecureFault | Unauthorised   |    system fatal error    |
```
s/Unauthorised/Unauthorized/  (American English spelling)

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+|      |            | access to      |                          |
+|      |            | secure region  |                          |
+|      |            | from ns space  |                          |
++------+------------+----------------+--------------------------+
```
Just for completeness, I'd add the rows for exceptions 8-10 and 13 with the exception name, and indicate in the remarks column that it's not handled by Zephyr (maybe ``Not handled``)

MaureenHelm replied:
I don't think this table should even be here.

minosgalanakis replied:
Should I remove the whole Interrupt Controller section including this table?

MaureenHelm replied:
Yes

pfalcon replied:
> I don't think this table should even be here.
> Should I remove the whole Interrupt Controller section including this table?

Note that we have  a similar case in #8408 where I also suggested to remove the whole "Interrupt Controller" section (which was  "TBD" there). We should give users detailed instruction on board bring-up with Zephyr (i.e. building/flashing), and just link to manuals for peculiar info like IRQ maps.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+
+Mapping from the ARM V2M Beetle Board pins to GPIO controllers:
+
+.. hlist::
```
Thank you for using an hlist directive!

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+============
+
+Musca has a built-in RGB LED connected to GPIO[4:2] pins.
+- Red LED connected at GPIO[2] pin,with optional PWM0.
```
add a blank line before the first item of a list.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+- Green LED connected at GPIO[3] pin,with optional PWM0.
+- Blue LED connected at GPIO[4] pin,with optional PWM0.
+
+Note
```
change this to:

    .. note::  The SCC registers select the functions of pins GPIO[4:2].

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+V2M Musca has a 32.768kHz crystal clock. The clock goes to a PLL and is
+multiplied to drive the Cortex-M33 processors and SSE-200 subsystem. The
+default is 50MHz but can be increased to 170MHz maximum for the secondary
+processor (CPU1). The maximum clock frequency for the primary processor
```
How... with a jumper?

minosgalanakis replied:
This section has been removed. Even though it is worded as such in the TRM, it is not advised to change the PLL frequencies for different core clock speeds.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+
+The ARM Musca processor has two UARTs. Both the UARTs have only two wires for
+RX/TX and no flow control (CTS/RTS) or FIFO. The Zephyr console output, by
+default, is utilizing UART1.
```
s/is utilizing/uses/

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+Security components
+===================
+
+- Implementation Defined Attribution Unit (IDAU).
```
You're assuming folks know what these are.  Maybe add a brief explanation, particularly for the acronyms.

minosgalanakis replied:
I have added hyperlinks to the official ARM Developer pages, created specifically to summarize their purpose.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+Serial Configuration Controller (SCC)
+=====================================
+
+The ARM Musca test chip implements Serial Configuration Control (SCC) register.
```
``implements a Serial ...``

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+=====================================
+
+The ARM Musca test chip implements Serial Configuration Control (SCC) register.
+The purpose of this register is to allow individual control of clocks,
```
Fix the punctuation with:

    The purpose of this register is to allow individual control of clocks,
    reset-signals and interrupts to peripherals, and pin-muxing.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+Musca supports the v8m security extension, and by default boots to the secure
+state.
+
+When building a secure(S)/non-secure(NS application, the S one will
```
Change to:

    When building a secure/non-secure application, the secure application will
    have to set the idau/sau and mpc configuration to permit access from the
    non-secure application before jumping.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+- Remote reset.
+
+For more details please refer
+to `DAPLink Website`_.
```
``to the``

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+
+You can build applications in the usual way. Here is an example for
+the :ref:`hello_world` application.
+
```
Looks like you left off the instruction steps

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+
+The process is similar to before, with the added configuration of setting-up
+the secure/non-secure configuration entry on zephyr.
+
```
Instruction steps missing here too?

minosgalanakis replied:
We need to decide on how to implement Zephyr S/NS builds and how it will be expressed in the cmake system. This section will have to be updated when it is ready.

agross-linaro replied:
we have basic secure only app instructions

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+Uploading an application to V2M Musca
+-------------------------------------
+
+In order to flash an application to V2M Musca, it needs to be converted to
```
For clarity:

    Applications must be converted to Intel's hex format before
    being flashed to a VM2 Musca board.

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+Intel's hex format. An optional bootloader can be prepended to the image.
+The QSPI flash base address alias is 0x200000.
+
+The image off-set is calculated by adding the flash off-set to the
```
``offset`` is one word without a hyphen

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+   # For a 256K bootloader IMAGE_OFFSET = $QSPI_FLASH_OFFSET + 0x40000
+   srec_cat $BIN_BOOLOADER -Binary -offset 0x200000 $BIN_SNS -Binary -offset 0x240000 -o $HEX_FLASHABLE -Intel
+
+Connect the V2M Musca to your host computer using the USB port and you should
```
split into two sentences:

    ... the USB port.  You should ...

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+   srec_cat $BIN_BOOLOADER -Binary -offset 0x200000 $BIN_SNS -Binary -offset 0x240000 -o $HEX_FLASHABLE -Intel
+
+Connect the V2M Musca to your host computer using the USB port and you should
+see a USB connection which exposes a Mass Storage (MBED) and a USB Serial Port.
```
change ``which exposes`` to ``exposing``

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+
+Connect the V2M Musca to your host computer using the USB port and you should
+see a USB connection which exposes a Mass Storage (MBED) and a USB Serial Port.
+Copy the generated zephyr.hex in the MBED drive.
```
add back ticks:  

    ``zephyr.hex``

dbkinder commented on boards/arm/v2m_musca/doc/v2m_musca.rst:
```diff
@@ -0,0 +1,412 @@
+Connect the V2M Musca to your host computer using the USB port and you should
+see a USB connection which exposes a Mass Storage (MBED) and a USB Serial Port.
+Copy the generated zephyr.hex in the MBED drive.
+Reset the board and you should be able to see on the corresponding Serial Port
```
Change to:

    Reset the board, and you should see the following message on the corresponding
    serial port:

MaureenHelm commented on boards/arm/v2m_musca/dts.fixup:
```diff
@@ -0,0 +1,82 @@
+/* SoC level DTS fixup file */
```
Add license/copyright header to all new files, not just .c/.h files.

ioannisg commented on boards/arm/v2m_musca/dts.fixup:
```diff
@@ -0,0 +1,82 @@
+/* SoC level DTS fixup file */
```
I am wondering, @galak : wouldn't it be more readable, if we had 2 different fixup files, for Secure and for Non-Secure hardware description?

ioannisg replied:
Actually: would it be possible?

ioannisg commented on boards/arm/v2m_musca/v2m_musca-nonsecure.dtsi:
```diff
@@ -0,0 +1,59 @@
+			label = "UART_1";
+		};
+
+		scc@4010c000 {
```
Doesn't need a "label"?

ioannisg commented on boards/arm/v2m_musca/v2m_musca-secure.dtsi:
```diff
@@ -0,0 +1,59 @@
+		reg = <0x10200000 0x800000>;
+	};
+
+	sram0: memory@30000000 {
```
Is this correct? SRAM starting at 0x3....?

karl-zh replied:
0x30000000 is the address in secure map for internal SRAM

ioannisg commented on boards/arm/v2m_musca/v2m_musca-secure.dtsi:
```diff
@@ -0,0 +1,59 @@
+		};
+
+		scc@5010c000 {
+			compatible = "arm,scc";
```
Same question as in the nonsecure .dtsi (for the missing label)

ioannisg commented on boards/arm/v2m_musca/v2m_musca.dts:
```diff
@@ -0,0 +1,36 @@
+#include <arm/armv8-m.dtsi>
+
+#if defined (CONFIG_ARM_NONSECURE_FIRMWARE)
+#include "v2m_musca-nonsecure.dtsi"
```
General comment here, about naming.
"<SOC name>-secure.dtsi is not so good, IMO, because I would like to preserve backwards compatibility with images without TrustZone consideration. So I would prefer that the "secure" .dtsi is the default .dtsi and, therefore, it should skip the "secure" from the name. And we should follow this convention for all ARMv8-M MCUs.

For Non-Secure: I would personally prefer to introduce a "_ns" suffix as a convention, to name the Non-Secure .dtsi headers. Unless you have strong preference on "-nonsecure" suffix, of course :)

ioannisg commented on boards/arm/v2m_musca/v2m_musca.dts:
```diff
@@ -0,0 +1,36 @@
+
+	chosen {
+		zephyr,console = &uart1;
+		zephyr,sram = &sram0;
```
This one will work, if sram0 is defined differently in the two .dtsi files. If not, we need to implement some kind of partition, here (in the Board definition) and choose the right sram for the secure and the non-secure image.

ioannisg reviewed (changes requested):
Pr looks OK. I left some comments, and some "open" issues to conclude on.

galak commented:
@minosgalanakis Can you address the doc feedback.

galak commented on drivers/serial/uart_pl011.c:
```diff
@@ -0,0 +1,512 @@
+static int pl011_poll_in(struct device *dev, unsigned char *c)
+{
+	if (!pl011_is_readable(dev))
+		return -1;
```
What errno, seems most drivers are return -1

MaureenHelm replied:
ok

galak commented on drivers/serial/uart_pl011.c:
```diff
@@ -0,0 +1,512 @@
+	 */
+	if((bauddiv < (1u << PL011_FBRD_WIDTH))
+		|| (bauddiv > (65535u << PL011_FBRD_WIDTH))) {
+		return 1;
```
fixed

galak commented on drivers/serial/uart_pl011.c:
```diff
@@ -0,0 +1,512 @@
+
+	/* Set baud rate */
+	ret = pl011_set_baudrate(dev, CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
+					DEV_DATA(dev)->baud_rate);
```
changed to check ret.

minosgalanakis commented:
> 
> 
> @minosgalanakis Can you address the doc feedback.

@galak .I have addressed the issues and pushed changes to my private fork.
https://github.com/minosgalanakis/zephyr/commit/a54ec8e9e21152416e232fe291232b8e661f148f

They can bundled with the rest of the project files, if acceptable.

galak commented:
@dbkinder Andy updated the docs, can you take a look again.

MaureenHelm reviewed (dismissed):
There are still a few new files (yaml, dtsi, CMakeLists.txt) missing a license/copyright header.

galak commented:
@dbkinder ping on docs review.

galak commented:
> There are still a few new files (yaml, dtsi, CMakeLists.txt) missing a license/copyright header.

should be fixed now.

dbkinder reviewed (approved):
LGTM, thanks!

galak commented:
> @galak .I have addressed the issues and pushed changes to my private fork.
> [minosgalanakis@a54ec8e](https://github.com/minosgalanakis/zephyr/commit/a54ec8e9e21152416e232fe291232b8e661f148f)
> 
> They can bundled with the rest of the project files, if acceptable.

Are these up to date with the doc fixes @agross-linaro made?

minosgalanakis commented:
> 
> 
> > @galak .I have addressed the issues and pushed changes to my private fork.
> > [minosgalanakis@a54ec8e](https://github.com/minosgalanakis/zephyr/commit/a54ec8e9e21152416e232fe291232b8e661f148f)
> > They can bundled with the rest of the project files, if acceptable.
> 
> Are these up to date with the doc fixes @agross-linaro made?

I have synced with  @agross-linaro changes and fixed a minor name issue s/Beetle/musca. 3871fd561e07c60630c65928b27f04246f8f8082

agross-linaro commented:
need to squash that before merging, unless we don't care about the patch count/naming.

SebastianBoe commented:
I was not able to get this to build with the latest Zephyr SDK (0.9.3).

I see there are no CI failuers, is CI covering this board? From the CI logs, it seems that this board is untested.

SebastianBoe commented:
recheck

ioannisg commented:
Shall we list this as a stale PR?

agross-linaro commented:
No we have updates.  Just haven't pushed them.  I'll ask kumar to do it again today.

galak commented:
@ioannisg @MaureenHelm can you guys re-review.

ioannisg commented on boards/arm/v2m_musca/Kconfig.board:
```diff
@@ -0,0 +1,14 @@
+config BOARD_MUSCA_A_NONSECURE
+	bool "ARM Cortex-M33 SMM on V2M-MUSCA (Non-Secure)"
+	depends on SOC_SERIES_MUSCA
+	select ARM_NONSECURE_FIRMWARE
```
The internal ARM option shall not be used directly, any more, since we, now, have top-level options for secure and non-secure firmware; `TRUSTED_EXECUTION_SECURE `and `TRUSTED_EXECUTION_NONSECURE`. These options can be selected in prj.conf of the user application. And the whole dependency framework for TrustZone-M and TEE is available, as well.

I don't think we need to define `BOARD_MUSCA_A_NONSECURE `option here.
If you really want to have a Non-Secure-only Board, you could just directly select `TRUSTED_EXECUTION_NONSECURE=y` in the Board defconfig.

ioannisg replied:
You could leave `BOARD_MUSCA_A_NONSECURE `as internal option (hidden) to implement your conditionals in Kconfig.defconfig, but it should be indirectly set by `TRUSTED_EXECUTION_NONSECURE`.

agross-linaro replied:
I removed this whole part.  It is unnecessary.

ioannisg commented on boards/arm/v2m_musca/v2m_musca_defconfig:
```diff
@@ -0,0 +1,24 @@
+CONFIG_BOARD_MUSCA_A=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_RUNTIME_NMI=y
+
```
No MPU support, @galak @agross-linaro ?

ioannisg replied:
You should also enable TrustZone-M: 
```
# Enable TrustZone-M
CONFIG_ARM_TRUSTZONE_M=y

```
So the user has option to use this board with or without support for TrustZone-M.

agross-linaro replied:
added this in

ioannisg commented on boards/arm/v2m_musca/v2m_musca_nonsecure_defconfig:
```diff
@@ -0,0 +1,24 @@
+CONFIG_SOC_FAMILY_ARM=y
+CONFIG_SOC_SERIES_MUSCA=y
+CONFIG_SOC_V2M_MUSCA_A=y
+CONFIG_BOARD_MUSCA_A_NONSECURE=y
```
See comment, above, in the definition of this option.

agross-linaro replied:
going to remove this file, per our discussion

agross-linaro replied:
I've found that I cannot remove this as the build system cannot adequately determine the DTS file.  We either have to leave the -DBOARD=v2m_musca_nonsecure as a build option and the only way to get a nonsecure image, or we come up with another way of determining the appropriate base DTS file to compile in the cmake scripts based off of some selection process.

ioannisg commented on boards/arm/v2m_musca/v2m_musca_nonsecure_defconfig:
```diff
@@ -0,0 +1,24 @@
+CONFIG_BOARD_MUSCA_A_NONSECURE=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_RUNTIME_NMI=y
+
```
MPU and TrustZone-M?

ioannisg commented on soc/arm/arm/musca_a/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+config SOC_V2M_MUSCA_A
+	bool "ARM Cortex-M33 SMM-SSE-200 on V2M-MUSCA"
+	select CPU_CORTEX_M33
+	select ARMV8_M_SE
```
You should not directly select ARMV8M_SE.
Instead, you need to select CPU_HAS_ARM_SAU, to indicate the TEE capabilities of the SOC.

agross-linaro replied:
ditto

ioannisg reviewed (changes requested):
I did a pass, focusing on TrustZone-M and TEE.

galak commented:
@ioannisg We pushed some updates from @agross-linaro, can you take a look again.

ioannisg commented on boards/arm/v2m_musca/v2m_musca.yaml:
```diff
@@ -0,0 +1,7 @@
+type: mcu
+arch: arm
+toolchain:
+  - gnuarmemb
```
@galak do we need to update the toolchain support?

ioannisg replied:
I mean, i suppose the SDK is supporting/will support ARMv8-M

galak replied:
I added xtools since that will be how we get v8m support.

agross-linaro replied:
yeah we will have it in the sdk ng.  This needs to be replaced with ds5 and arm gcc.  I'll make that change.

ioannisg replied:
cool

agross-linaro replied:
added in ds5 (external) line underneath xtools

ioannisg commented on boards/arm/v2m_musca/v2m_musca_defconfig:
```diff
@@ -0,0 +1,27 @@
+CONFIG_SOC_V2M_MUSCA_A=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_RUNTIME_NMI=y
+CONFIG_TRUSTED_EXECUTION_SECURE=y
```
My only concern, here is: what if the user wants to use the Musca board as a legacy (no TEE) support?
Why is the TRUSTED_EXECUTION_SECURE set by default? Shouldn't this be better set on the project level? I don't really mind, actually; I am just throwing the idea.

agross-linaro replied:
well, i think we were using two different defconfigs to determine secure or non-secure.  A non-TEE system would be running secure with SDAU/IDAU configured to fully open, correct?

agross-linaro replied:
we can remove the TEE from secure, but it has to remain for non-secure because otherwise it is not a sane config.  Then it's up to the application to define whether they want a TEE secure image or a v7m compatible.

agross-linaro replied:
Ok I see your point.  We just have to punt this to the application.  This means each possible combination of an application (secure, nonsecure, v7m compat) has to be handled individually in that application's config file properly.  This will make reusing current apps troublesome.

ioannisg commented on soc/arm/arm/musca_a/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+config SOC_V2M_MUSCA_A
+	bool "ARM Cortex-M33 SMM-SSE-200 on V2M-MUSCA"
+	select CPU_CORTEX_M33
+	select CPU_HAS_ARM_SAU
```
@galak @agross-linaro do you select, somewhere CPU_HAS_ARM_MPU? Have I missed it?

agross-linaro replied:
I swear I added that.  But perhaps I lost it when I shifted things around.  Let me add it back.

ioannisg replied:
Sure :)

agross-linaro replied:
done

ioannisg commented on boards/arm/v2m_musca/v2m_musca-common.dtsi:
```diff
@@ -0,0 +1,51 @@
+/*
```
@galak @agross-linaro so, it seems to me that this file is included by _both_ the secure and the non-secure dts files.
Just to double-check; the reg properties (and corresponding Register Base Addresses) are the _same_ for Secure and Non-Secure access?? If so, this should work, otherwise, no.

And then, I have, again, the earlier comment; shouldn't this file be under ./dts/arm? Cause it is not really board specific.

agross-linaro replied:
it may or may not be.  It depends on how the SOC vendors do their base board support and board specific add-ons.  In the case of this specific board, it's a one off board with a specific implementation.  So i suppose it could go in dts/arm, but it'll only be used for 1 board.  The v2m_musca-a.  If the musca-b is the same, then there is an argument for dts/arm.  But we haven't seen the memory map.

agross-linaro replied:
this is required due to the ranges property being used to steer the addressing of the peripherals.

Commit:
arch: arm: mpu: arm_mpu: Fix compiler error on stack_obj

This patch adds a required cast for the size calculation utilizing
pointer arithmetic.  The stack_obj needs a cast to u32_t.  This was
caught using a newer compiler.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

galak commented:
@ioannisg pushed an update, can you review again.

[ There's a MPU build fix that you guys might also need ]

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -207,7 +207,7 @@ void arm_core_mpu_configure_user_context(struct k_thread *thread)
 	 * MPU faults.
 	 */
-	size += thread->stack_info.start - thread->stack_obj;
+	size += thread->stack_info.start - (u32_t)thread->stack_obj;
```
@galak nice! We had fixed that downstream, but deferred from fixing this upstream until we actually get a v8m board with USERSPACE support :) No objection of fixing it under this PR.

Could you, perhaps, replace `(u32_t)thread->stack_obj` with `base`?

galak replied:
Done

agross-linaro replied:
yeah we'll do that

ioannisg commented on soc/arm/arm/musca_a/arm_mpu_regions.c:
```diff
@@ -0,0 +1,26 @@
+			 REGION_FLASH_ATTR(CONFIG_FLASH_BASE_ADDRESS,
+					   CONFIG_FLASH_SIZE*1024)},
+	/* Region 1 */
+	{CONFIG_SRAM_BASE_ADDRESS,
```
minor: leave a space here, to align with line 12, or remove the space from line 12 :)

galak replied:
Done, reworked formating

ioannisg commented on soc/arm/arm/musca_a/arm_mpu_regions.c:
```diff
@@ -0,0 +1,26 @@
+					   CONFIG_FLASH_SIZE*1024)},
+	/* Region 1 */
+	{CONFIG_SRAM_BASE_ADDRESS,
+			 "RAM_0",
```
That should be "SRAM_0"

galak replied:
Done
